File: workspace/2023-04-05/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fn n -- fn )
    dup 0 = if
        drop        ( fn n -- fn )
    else
        tuck *      ( fn n -- n fn+n )
        swap        ( n fn+n -- fn+n n )
        1 -         ( fn+n n -- fn+n n-1 )
        _factorial  ( fn+n n-1 -- fn )
    endif
;

: factorial ( n -- fn )
    1 swap      ( n -- 1 n )
    _factorial  ( 1 n -- fn )
;

: _show_factorials recursive ( max min -- )
    over over < if
        drop drop           ( max min -- )
    else
        dup factorial .
        1 +                 ( max min -- max min+1 )
        _show_factorials    ( max min+1 -- )
    endif
;

: show_factorials ( n -- )
    1
    _show_factorials
;

File: workspace/2023/2023-03-20/try-erlang.erlang/factorial.erl

main(_) ->
    show_factorials(1, 10).

show_factorials(Min, Max) when Min > Max ->
    ok;
show_factorials(Min, Max) ->
    io:format("factorial(~w) = ~w~n", [Min, factorial(Min)]),
    show_factorials(Min + 1, Max).

factorial(0) ->
    1;
factorial(N) ->
    N * factorial(N - 1).

